lib/SILOptimizer/Analysis/EscapeAnalysis.cpp

```diff
@@ -575,32 +827,111 @@ getUsePoints(CGNode *Node, llvm::SmallVectorImpl<SILNode *> &UsePoints) {
-      clearWorkListFlags(WorkList);
-      return true;
+// Traverse backward from startNode and return true if \p visitor did not halt
+// traversal..
+//
+// The graph may have cycles.
+template <typename CGPredVisitor>
+bool EscapeAnalysis::ConnectionGraph::backwardTraverse(
+    CGNode *startNode, CGPredVisitor &&visitor) {
+  CGNodeWorklist worklist(this);
+  worklist.push(startNode);
+
+  for (unsigned idx = 0; idx < worklist.size(); ++idx) {
+    CGNode *reachingNode = worklist[idx];
+
+    for (Predecessor pred : reachingNode->Preds) {
+      switch (visitor(pred)) {
+      case Traversal::Follow: {
+        CGNode *predNode = pred.getPointer();
+        worklist.tryPush(predNode);
+        break;
+      }
+      case Traversal::Backtrack:
+        break;
+      case Traversal::Halt:
+        return false;
+      }
     }
-    for (Predecessor Pred : Reachable->Preds) {
-      CGNode *PredNode = Pred.getPointer();
-      if (!PredNode->isInWorkList) {
-        PredNode->isInWorkList = true;
-        WorkList.push_back(PredNode);
+  }
+  return true;
+}
+
+// Traverse forward from startNode, following defer edges and return true if \p
+// visitor did not halt traversal.
+//
+// The graph may have cycles.
+template <typename CGNodeVisitor>
+bool EscapeAnalysis::ConnectionGraph::forwardTraverseDefer(
+    CGNode *startNode, CGNodeVisitor &&visitor) {
+  CGNodeWorklist worklist(this);
+  worklist.push(startNode);
+
+  for (unsigned idx = 0; idx < worklist.size(); ++idx) {
+    CGNode *reachableNode = worklist[idx];
+
+    for (CGNode *deferNode : reachableNode->defersTo) {
+      switch (visitor(deferNode)) {
+      case Traversal::Follow:
+        worklist.tryPush(deferNode);
+        break;
+      case Traversal::Backtrack:
+        break;
+      case Traversal::Halt:
+        return false;
       }
     }
   }
-  clearWorkListFlags(WorkList);
-  return false;
+  return true;
 }
 
+// Follow transitive pointsTo edges from \p startNode.
+// This path may not contain cycles.
+template <typename CGNodeVisitor>
+bool EscapeAnalysis::ConnectionGraph::forwardTraversePointsToEdges(
+    CGNode *startNode, CGNodeVisitor &&visitor) {
+  CGNodeWorklist visitedNodes(this);
+  CGNode *pointer = startNode;
+  while ((pointer = pointer->getPointsToEdge())) {
+    if (!visitedNodes.tryPush(pointer))
+      return true;
+
+    if (!visitor(pointer)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Returns the last content node in the chain of pointsTo edges pointed to by
+// this node.
+//
+// If this node has no content, return the node itself. If has content or is
+// already a content follow the chain of pointsTo edges and return the last
+// content node in the chain.
+CGNode *EscapeAnalysis::ConnectionGraph::getPointsToEnd(CGNode *node) {
+  CGNode *content = node->isContent() ? node : node->getContentNodeOrNull();
+  if (!content)
+    return node;
+  forwardTraversePointsToEdges(content, [&](CGNode *pointsTo) {
```

Comment:
This is an obscure way to formulate the algorithm. It took me a while to understand it.
It would be way easier to not use forwardTraversePointsToEdges.